README.md
![HHTabBarView](https://github.com/Tobaloidee/HHTabBarView/blob/master/logo/logotype-a.png)

A lightweight customized tabbar view.

[![Build Status](https://travis-ci.org/hemangshah/HHTabBarView.svg?branch=master)](https://travis-ci.org/hemangshah/HHTabBarView)
[![License](https://img.shields.io/badge/License-MIT-lightgrey.svg)](https://github.com/hemangshah/HHTabBarView/blob/master/LICENSE)
[![Platform](https://img.shields.io/badge/Platforms-iOS-red.svg)](https://www.apple.com/in/ios/)
[![Swift 4.x](https://img.shields.io/badge/Swift-4.x-blue.svg)](https://swift.org/)
[![MadeWithLove](https://img.shields.io/badge/Made%20with%20%E2%9D%A4-India-green.svg)](https://madewithlove.org.in/)
[![Awesome-Swift](https://cdn.rawgit.com/sindresorhus/awesome/d7305f38d29fed78fa85652e3a63e154dd8e8829/media/badge.svg)](https://github.com/matteocrippa/awesome-swift/)

1. [Screenshots](#screenshots)
2. [Features](#features)
3. [Installation](#installation)
4. [Setup](#setup)
5. [ToDos](#todos)
6. [Credits](#credits)
7. [Thanks](#thank-you)
8. [License](#license)

## Screenshots

<table>
<tr>
<td align="center"><img src = "https://github.com/hemangshah/HHTabBarView/blob/master/Screenshots/1.png" alt = "Only Icon"></td>
<td align="center"><img src = "https://github.com/hemangshah/HHTabBarView/blob/master/Screenshots/2.png" alt = "Icon (top) with title (bottom)"></td>
<td align="center"><img src = "https://github.com/hemangshah/HHTabBarView/blob/master/Screenshots/3.png" alt = "Only Title"></td>
</tr>
<tr>
<td align="center"><img src = "https://github.com/hemangshah/HHTabBarView/blob/master/Screenshots/4.png" alt = "Icon (left) with title (right)"></td>
<td align="center"><img src = "https://github.com/hemangshah/HHTabBarView/blob/master/Screenshots/5.png" alt = "Icon (right) with title (left)"></td>
<td align="center"><img src = "https://github.com/hemangshah/HHTabBarView/blob/master/Screenshots/6.png" alt = "Icon (bottom) with title (top)"></td>
</tr>
<tr>
<td align="center"><img src = "https://github.com/hemangshah/HHTabBarView/blob/master/Screenshots/7.png" alt = "HHTabBarView top position"></td>
<td align="center"><img src = "https://github.com/hemangshah/HHTabBarView/blob/master/Screenshots/8.png" alt = "HHTabBarView top position (iPhoneX)"></td>
</tr>
<tr>
<td colspan="3" align="center"><img src = "https://github.com/hemangshah/HHTabBarView/blob/master/Screenshots/HHTabBarFlow.gif" alt = "Usage"></td>
</tr>
</table>

## Features

1. Easily Configurable and Setup. Create tabs with Title, or Image or both. 
2. Dynamic Tabs Configurations.
3. Detect Taps in a completion block.
4. Show/Hide Badge Value in individual tabs. Easily Configure as per the needs.
5. Lock/Unlock particular tabs.
6. Easily show/hide UINavigationBar and HHTabBarView.
7. Lightweight with zero dependancies.
8. Change UI of HHTabBarView (LeftToRight or RightToLeft) as per the needs.
9. Change HHTabBarView position as per UI requirements. Supports: Top and Bottom (Default). 

##### Note: I don't recommended displaying HHTabBarView at the top. It's always good at the bottom.

## Installation

1. **Manually** – Add `HHTabBarView/Source` folder to your Project. And you're good to use `HHTabBarView`.

2. **CocoaPods**: – `pod 'HHTabBarView'`
    
> You can read the [CHANGELOG](https://github.com/hemangshah/HHTabBarView/blob/master/CHANGELOG.md) file for a particular release.

## Setup

**Important**: Please note that `HHTabBarView` is currently not supports `UIStoryBoard`. Means, you will have to create `HHTabBarView` programmatically. It is advised to setup `HHTabBarView` in `AppDelegate.swift` for your easyness.

1.  Initialize and keeping reference of `HHTabBarView`. 📌
````swift
    let hhTabBarView = HHTabBarView.shared
````

2.  Keeping reference of iOS default `UITabBarController`. 📌
````swift
    let referenceTabBarController = HHTabBarView.shared.referenceUITabBarController
````
    
3. Setup referenced `UITabBarController` 📌
````swift
    func setupReferenceUITabBarController() -> Void {
        
        //Creating a storyboard reference
        let storyboard = UIStoryboard.init(name: "Main", bundle: Bundle.main)
        
        //Creating navigation controller for navigation inside the first tab.
        let navigationController1: UINavigationController = UINavigationController.init(rootViewController: storyboard.instantiateViewController(withIdentifier: "FirstViewControllerID"))
        
        //Creating navigation controller for navigation inside the second tab.
        let navigationController2: UINavigationController = UINavigationController.init(rootViewController: storyboard.instantiateViewController(withIdentifier: "SecondViewControllerID"))
        
        //Update referenced TabbarController with your viewcontrollers
        referenceTabBarController.setViewControllers([navigationController1, navigationController2], animated: false)
    }
````    
    
4. Setup `HHTabBarView` 📌
````swift
    //Update HHTabBarView reference with the tabs requires.
    func setupHHTabBarView() -> Void {
        
        //Default & Selected Background Color
        let defaultTabColor = UIColor.white
        let selectedTabColor = UIColor.init(red: 234/255, green: 218/255, blue: 195/255, alpha: 1.0)
        let tabFont = UIFont.init(name: "Helvetica-Light", size: 14.0)
        
        //Create Custom Tabs
        let t1 = HHTabButton.init(withTitle: "Calendar", tabImage: UIImage.init(named: "Calendar")!, index: 0)
        t1.titleLabel?.font = tabFont
        t1.titleLabel?.textColor = UIColor.black
        t1.setHHTabBackgroundColor(color: defaultTabColor, forState: .normal)
        t1.setHHTabBackgroundColor(color: selectedTabColor, forState: .selected)
        
        let t2 = HHTabButton.init(withTitle: "Refresh", tabImage: UIImage.init(named: "Refresh")!, index: 1)
        t2.titleLabel?.font = tabFont
        t2.titleLabel?.textColor = UIColor.black
        t2.setHHTabBackgroundColor(color: defaultTabColor, forState: .normal)
        t2.setHHTabBackgroundColor(color: selectedTabColor, forState: .selected)
        
        //Note: As HHTabButton are subclassed of UIButton so you can modify it as much as possible.
        
        //Set Custom Tabs
        hhTabBarView.tabBarTabs = [t1, t2]
        
        //Set Default Index for HHTabBarView.
        hhTabBarView.defaultIndex = 1
        
        //Show Animation on Switching Tabs
        hhTabBarView.tabChangeAnimationType = .none
        
        //Handle Tab Change Event
        hhTabBarView.onTabTapped = { (tabIndex) in
            print("Selected Tab Index:\(tabIndex)")
        }
    }
````

5. Setup `window` of your application inside the 📌
````swift
    func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplicationLaunchOptionsKey: Any]?) -> Bool {
        // Override point for customization after application launch.
        
        //Setup HHTabBarView
        setupReferenceUITabBarController()
        setupHHTabBarView()
        
        //Setup Application Window
        self.window = UIWindow.init(frame: UIScreen.main.bounds)
        self.window?.rootViewController = self.referenceTabBarController
        self.window?.makeKeyAndVisible()
        
        return true
    }
````

6. Done! ✅    

<hr>

## ToDo[s]

- [x] Update README with multiple example usage and screenshots.

You can [watch](https://github.com/hemangshah/HHTabBarView/subscription) to <b>HHTabBarView</b> to see continuous updates. Stay tuned.

<b>Have an idea for improvements of this class?
Please open an [issue](https://github.com/hemangshah/HHTabBarView/issues/new).</b>
    
## Credits

<b>[Hemang Shah](https://about.me/hemang.shah)</b>

**You can shoot me an [email](http://www.google.com/recaptcha/mailhide/d?k=01IzGihUsyfigse2G9z80rBw==&c=vU7vyAaau8BctOAIJFwHVbKfgtIqQ4QLJaL73yhnB3k=) to contact.**
   
## Thank You!!

See the [contributions](https://github.com/hemangshah/HHTabBarView/blob/master/CONTRIBUTIONS.md) for details.

## License

The MIT License (MIT)

> Read the [LICENSE](https://github.com/hemangshah/HHTabBarView/blob/master/LICENSE) file for details.


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.

Latest release is available [here](https://github.com/hemangshah/HHTabBarView/releases/latest).

## [2.3.0](https://github.com/hemangshah/HHTabBarView/releases/tag/2.3.0) – Sep 24, 2019
- Fixed: We can now go back by tapping on the same tab again in navigation hierarchy.
- Support for Swift 5.0

## [2.2.0](https://github.com/hemangshah/HHTabBarView/releases/tag/2.2.0) – Dec 15, 2018
- Fixed: https://github.com/hemangshah/HHTabBarView/issues/13
- Fixed: https://github.com/hemangshah/HHTabBarView/issues/12
- Support for Swift 4.2

## [2.1.0](https://github.com/hemangshah/HHTabBarView/releases/tag/2.1.0) – Sep 16, 2018
- Fixed: https://github.com/hemangshah/HHTabBarView/issues/9
- Minor code refactoring.

## [2.0.0](https://github.com/hemangshah/HHTabBarView/releases/tag/2.0.0) – Aug 18, 2018
- Updated code base to allow user to add HHTabBarView at top position as well.
- Minor code refactoring.

## [1.3.0](https://github.com/hemangshah/HHTabBarView/releases/tag/1.3.0) - Apr 02, 2018
- Left to Right and Right to Left UI management as per the language. Example: English <–> Arabic.
- Code refactoring.

## [1.2.0](https://github.com/hemangshah/HHTabBarView/releases/tag/1.2.0) - Feb 26, 2018
- Huge code refactoring.
- Swift 4.x support. Note: Revoked Swift 3.x support.
- Customized position for title and image.

## [1.1.2](https://github.com/hemangshah/HHTabBarView/releases/tag/1.1.2) - Nov 26, 2017
- Removed default black color for the HHTabButton title.
- Allows selecting a particular tab programmatically.
- Fixed a glitch when setting a title and an image together.

## [1.1.1](https://github.com/hemangshah/HHTabBarView/releases/tag/1.1.1) - Nov 07, 2017
- Updated for iPhone X.

## [1.1.0](https://github.com/hemangshah/HHTabBarView/releases/tag/1.1.0) - Oct 26, 2017
- Improvements.

## [1.0.0](https://github.com/hemangshah/HHTabBarView/releases/tag/1.0.0) - Oct 22, 2017
- First Release over CocoaPods.


HHTabBarView/HHTabBarView/Source/HHTabLabel.swift
//
//  HHTabLabel.swift
//  HHTabBarView
//
//  Created by Hemang Shah on 7/26/17.
//  Copyright © 2017 Hemang Shah. All rights reserved.
//

import UIKit

///This is to show badge inside the HHTabButton (i.e. HHTabBarView's tab). As it's subclassed of UILabel you can configure it as much as iOS supports.
public class HHTabLabel: UILabel {
    override init(frame: CGRect) {
        super.init(frame: frame)
        backgroundColor = .red
        layer.cornerRadius = frame.size.width/2.0
        layer.masksToBounds = true
        textColor = .white
        textAlignment = .center
        font = UIFont.init(name: "Helvetica-Light", size: 10.0)
        adjustsFontSizeToFitWidth = true
    }
    
    required public init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
}


HHTabBarView/HHTabBarView/Source/HHTabButton.swift
//
//  HHTabButton.swift
//  HHTabBarView
//
//  Created by Hemang Shah on 7/17/17.
//  Copyright © 2017 Hemang Shah. All rights reserved.
//

/*
 Idea of custom positions of title and image.
 Credits: https://gist.github.com/gbitaudeau/6540847de8f5ee9f2e0393a00d2cb11e
*/

import UIKit

///This is to show tabs inside the HHTabBarView. As it's subclassed of UIButton you can configure it as much as iOS supports.
public class HHTabButton: UIButton {
    
    ///Vertical Alignments
    public enum VerticalAlignment: String {
        case center, top, bottom, unset
    }
    
    ///Horizontal Alignments
    public enum HorizontalAlignment: String {
        case center, left, right, unset
    }

    ///Spacing between then image and title.
    public var imageToTitleSpacing: CGFloat = 8.0 {
        didSet {
            setNeedsLayout()
        }
    }
    
    ///Vertical alignment for the image.
    public var imageVerticalAlignment: VerticalAlignment = .unset {
        didSet {
            setNeedsLayout()
        }
    }
    
    ///Horizontal alignment for the image.
    public var imageHorizontalAlignment: HorizontalAlignment = .unset {
        didSet {
            setNeedsLayout()
        }
    }
    
    ///Set extra content edgeInsets.
    public var extraContentEdgeInsets: UIEdgeInsets = UIEdgeInsets.zero
    
    override public var contentEdgeInsets: UIEdgeInsets {
        get {
            return super.contentEdgeInsets
        }
        set {
            super.contentEdgeInsets = newValue
            self.extraContentEdgeInsets = newValue
        }
    }
    
    ///Set extra image edgeInsets.
    public var extraImageEdgeInsets: UIEdgeInsets = UIEdgeInsets.zero
    
    override public var imageEdgeInsets: UIEdgeInsets {
        get {
            return super.imageEdgeInsets
        }
        set {
            super.imageEdgeInsets = newValue
            self.extraImageEdgeInsets = newValue
        }
    }
    
    ///Set extra title edgeInsets.
    public var extraTitleEdgeInsets:UIEdgeInsets = UIEdgeInsets.zero
    
    override public var titleEdgeInsets: UIEdgeInsets {
        get {
            return super.titleEdgeInsets
        }
        set {
            super.titleEdgeInsets = newValue
            self.extraTitleEdgeInsets = newValue
        }
    }
    
    ///Unique index of tabs. (Should be start with 0)
    public var tabIndex:Int = 0
    
    ///To set badge value.
    internal var badgeValue: Int = 0 {
        willSet {
            self.addBadgeView()
        } didSet {
            self.updateBadgeView()
        }
    }

    ///Configure badge Label. Should be configure only after setting tabs array to HHTabBarView.
    public var badgeLabel: HHTabLabel?
    
    // MARK:Init
    required convenience public init(withTitle tabTitle: String?, tabImage: UIImage?, index: Int) {
        self.init(frame: CGRect.zero)
        self.setupButton(withTitle: tabTitle, tabImage: tabImage, index: index)
    }
    
    fileprivate func setupButton(withTitle tabTitle: String?, tabImage: UIImage?, index: Int) {
        self.backgroundColor = UIColor.clear
        self.autoresizingMask = [.flexibleWidth, .flexibleLeftMargin, .flexibleRightMargin, .flexibleHeight]
        self.tabIndex = index
        self.setImage(tabImage, for: .normal)
        self.setImage(tabImage, for: .selected)
        self.setTitle(tabTitle, for: .normal)
        self.setTitle(tabTitle, for: .selected)
    }
    
    //This is to add BadgeView within the HHTabButton. See badgeValue var to understand the usage.
    fileprivate func addBadgeView() {
        
        //If badge label is already created, no need to recreate.
        guard self.badgeLabel == nil else {
            return
        }
        
        let badgeSize: CGFloat = 20.0
        let margin: CGFloat = 3.0
        self.badgeLabel = HHTabLabel.init(frame: CGRect.init(x: self.frame.size.width - (badgeSize + margin), y: margin, width: badgeSize, height: badgeSize))
        self.badgeLabel!.isHidden = true //By default.
        self.addSubview(self.badgeLabel!)
    }
    
    //This is to update BadgeView within the HHTabButton. See badgeValue var to understand the usage.
    fileprivate func updateBadgeView() {
        if let label = self.badgeLabel {
            if self.badgeValue <= 0 {
                label.isHidden = true
                label.text = ""
            } else {
                label.isHidden = false
                label.text = String(self.badgeValue)
            }
        }
    }
    
    // MARK: Init
    public override init(frame: CGRect) {
        super.init(frame: frame)
    }
    
    required public init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    // MARK: LayoutSubviews
    override public func layoutSubviews() {
        if let imageSize = self.imageView?.image?.size,
            let font = self.titleLabel?.font,
            let textSize = self.titleLabel?.attributedText?.size() ?? self.titleLabel?.text?.size(withAttributes: [NSAttributedString.Key.font: font]) {
            
            var _imageEdgeInsets = UIEdgeInsets.zero
            var _titleEdgeInsets = UIEdgeInsets.zero
            var _contentEdgeInsets = UIEdgeInsets.zero
            
            let halfImageToTitleSpacing = imageToTitleSpacing / 2.0
            
            switch imageVerticalAlignment {
            case .bottom:
                _imageEdgeInsets.top = (textSize.height + imageToTitleSpacing) / 2.0
                _imageEdgeInsets.bottom = (-textSize.height - imageToTitleSpacing) / 2.0
                _titleEdgeInsets.top = (-imageSize.height - imageToTitleSpacing) / 2.0
                _titleEdgeInsets.bottom = (imageSize.height + imageToTitleSpacing) / 2.0
                _contentEdgeInsets.top = (min (imageSize.height, textSize.height) + imageToTitleSpacing) / 2.0
                _contentEdgeInsets.bottom = (min (imageSize.height, textSize.height) + imageToTitleSpacing) / 2.0
                //only works with contentVerticalAlignment = .center
                contentVerticalAlignment = .center
            case .top:
                _imageEdgeInsets.top = (-textSize.height - imageToTitleSpacing) / 2.0
                _imageEdgeInsets.bottom = (textSize.height + imageToTitleSpacing) / 2.0
                _titleEdgeInsets.top = (imageSize.height + imageToTitleSpacing) / 2.0
                _titleEdgeInsets.bottom = (-imageSize.height - imageToTitleSpacing) / 2.0
                _contentEdgeInsets.top = (min (imageSize.height, textSize.height) + imageToTitleSpacing) / 2.0
                _contentEdgeInsets.bottom = (min (imageSize.height, textSize.height) + imageToTitleSpacing) / 2.0
                //only works with contentVerticalAlignment = .center
                contentVerticalAlignment = .center
            case .center:
                //only works with contentVerticalAlignment = .center
                contentVerticalAlignment = .center
                break
            case .unset:
                break
            }
            
            switch imageHorizontalAlignment {
            case .left:
                _imageEdgeInsets.left = -halfImageToTitleSpacing
                _imageEdgeInsets.right = halfImageToTitleSpacing
                _titleEdgeInsets.left = halfImageToTitleSpacing
                _titleEdgeInsets.right = -halfImageToTitleSpacing
                _contentEdgeInsets.left = halfImageToTitleSpacing
                _contentEdgeInsets.right = halfImageToTitleSpacing
            case .right:
                _imageEdgeInsets.left = textSize.width + halfImageToTitleSpacing
                _imageEdgeInsets.right = -textSize.width - halfImageToTitleSpacing
                _titleEdgeInsets.left = -imageSize.width - halfImageToTitleSpacing
                _titleEdgeInsets.right = imageSize.width + halfImageToTitleSpacing
                _contentEdgeInsets.left = halfImageToTitleSpacing
                _contentEdgeInsets.right = halfImageToTitleSpacing
            case .center:
                _imageEdgeInsets.left = textSize.width / 2.0
                _imageEdgeInsets.right = -textSize.width / 2.0
                _titleEdgeInsets.left = -imageSize.width / 2.0
                _titleEdgeInsets.right = imageSize.width / 2.0
                _contentEdgeInsets.left = -((imageSize.width + textSize.width) - max (imageSize.width, textSize.width)) / 2.0
                _contentEdgeInsets.right = -((imageSize.width + textSize.width) - max (imageSize.width, textSize.width)) / 2.0
            case .unset:
                break
            }
            
            _contentEdgeInsets.top += extraContentEdgeInsets.top
            _contentEdgeInsets.bottom += extraContentEdgeInsets.bottom
            _contentEdgeInsets.left += extraContentEdgeInsets.left
            _contentEdgeInsets.right += extraContentEdgeInsets.right
            
            _imageEdgeInsets.top += extraImageEdgeInsets.top
            _imageEdgeInsets.bottom += extraImageEdgeInsets.bottom
            _imageEdgeInsets.left += extraImageEdgeInsets.left
            _imageEdgeInsets.right += extraImageEdgeInsets.right
            
            _titleEdgeInsets.top += extraTitleEdgeInsets.top
            _titleEdgeInsets.bottom += extraTitleEdgeInsets.bottom
            _titleEdgeInsets.left += extraTitleEdgeInsets.left
            _titleEdgeInsets.right += extraTitleEdgeInsets.right
            
            super.imageEdgeInsets = _imageEdgeInsets
            super.titleEdgeInsets = _titleEdgeInsets
            super.contentEdgeInsets = _contentEdgeInsets
            
        } else {
            super.imageEdgeInsets = extraImageEdgeInsets
            super.titleEdgeInsets = extraTitleEdgeInsets
            super.contentEdgeInsets = extraContentEdgeInsets
        }
        
        super.layoutSubviews()
    }
}


HHTabBarView/HHTabBarView/FirstViewController.swift
//
//  FirstViewController.swift
//  HHTabBarView
//
//  Created by Hemang Shah on 7/17/17.
//  Copyright © 2017 Hemang Shah. All rights reserved.
//

import UIKit

class FirstViewController: UIViewController {

    override func viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view.
        self.title = "Location"
    }
    
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        // This will be helpful in locking the current tab if we're already in the first view-controller in navigation flow.
        HHTabBarView.shared.lockCurrentTab()
    }
    
    override func viewWillDisappear(_ animated: Bool) {
        super.viewWillDisappear(animated)
        // This will be helpful to unlock the locked tab when we're going forward from the first view-controller in navigation flow.
        HHTabBarView.shared.unlockAllTabs()
    }
}

extension FirstViewController: UITableViewDataSource {
    func numberOfSections(in tableView: UITableView) -> Int {
        return 1
    }
    
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return 100
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCell(withIdentifier: "cellIdentifier", for: indexPath)
        cell.textLabel?.text = "Location \(indexPath.row)"
        return cell
    }
}

extension FirstViewController: UITableViewDelegate {
    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        tableView.deselectRow(at: indexPath, animated: false)
        let viewController = UIViewController()
        viewController.title = "Location \(indexPath.row)"
        viewController.view.backgroundColor = UIColor.white
        self.navigationController?.pushViewController(viewController, animated: true)
    }
}


HHTabBarView/HHTabBarView/SecondViewController.swift
//
//  ViewController.swift
//  HHTabBarView
//
//  Created by Hemang Shah on 7/17/17.
//  Copyright © 2017 Hemang Shah. All rights reserved.
//

import UIKit

class SecondViewController: UIViewController {
    
    var apd = UIApplication.shared.delegate as! AppDelegate
    
    override func viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view, typically from a nib.
        self.title = "Refresh"
    }
    
    @IBAction func actionShowHideTabBar(_ sender: UISwitch) {
        apd.hhTabBarView.toggleShowOrHide()
    }
    
    @IBAction func actionShowHideNavigationBar(_ sender: UISwitch) {
        self.navigationController?.isNavigationBarHidden = !sender.isOn
    }
    
    @IBAction func actionShowHideBadge(_ sender: UISwitch) {
        if sender.isOn {
            apd.hhTabBarView.updateBadge(forTabIndex: 0, withValue: Int(arc4random()%30) + 1)
            apd.hhTabBarView.updateBadge(forTabIndex: 2, withValue: Int(arc4random()%30) + 1)
            apd.hhTabBarView.updateBadge(forTabIndex: 3, withValue: Int(arc4random()%30) + 1)
            apd.hhTabBarView.updateBadge(forTabIndex: 4, withValue: Int(arc4random()%30) + 1)
        } else {
            //Setting 0 will hide the Badge.
            apd.hhTabBarView.updateBadge(forTabIndex: 0, withValue: 0)
            apd.hhTabBarView.updateBadge(forTabIndex: 2, withValue: 0)
            apd.hhTabBarView.updateBadge(forTabIndex: 3, withValue: 0)
            apd.hhTabBarView.updateBadge(forTabIndex: 4, withValue: 0)
        }
    }
    
    @IBAction func actionLockUnlockLastTab(_ sender: UISwitch) {
        if sender.isOn {
            apd.hhTabBarView.lockTabIndexes = [4]
        } else {
            apd.hhTabBarView.lockTabIndexes = []
        }
    }
    
    @IBAction func actionShowTabChangeAnimation(_ sender: UISwitch) {
        if sender.isOn {
            apd.hhTabBarView.tabChangeAnimationType = .shake
        } else {
            apd.hhTabBarView.tabChangeAnimationType = .none
        }
    }
    
    @IBAction func actionReverseTabs(_ sender: UISwitch) {
        if sender.isOn {
            self.apd.hhTabBarView.rightToLeft()
        } else {
            self.apd.hhTabBarView.leftToRight()
        }
    }
}


HHTabBarView/HHTabBarView/Source/HHButton+Extension.swift
//
//  HHButton+Extension.swift
//  HHTabBarView
//
//  Created by Hemang on 03/10/18.
//  Copyright © 2018 Hemang Shah. All rights reserved.
//

import UIKit

public extension HHTabButton {
    
    internal func pulsate() {
        let pulse = CASpringAnimation(keyPath: "transform.scale")
        pulse.duration = 0.15
        pulse.fromValue = 0.95
        pulse.toValue = 1.0
        pulse.autoreverses = true
        pulse.repeatCount = 2
        pulse.initialVelocity = 0.5
        pulse.damping = 1.0
        layer.add(pulse, forKey: "pulse")
    }
    
    internal func flash() {
        let flash = CABasicAnimation(keyPath: "opacity")
        flash.duration = 0.15
        flash.fromValue = 1
        flash.toValue = 0.1
        flash.timingFunction = CAMediaTimingFunction(name: CAMediaTimingFunctionName.easeInEaseOut)
        flash.autoreverses = true
        flash.repeatCount = 3
        layer.add(flash, forKey: nil)
    }
    
    internal func shake() {
        let shake = CABasicAnimation(keyPath: "position")
        shake.duration = 0.05
        shake.repeatCount = 2
        shake.autoreverses = true
        let fromPoint = CGPoint(x: center.x - 5, y: center.y)
        let fromValue = NSValue(cgPoint: fromPoint)
        let toPoint = CGPoint(x: center.x + 5, y: center.y)
        let toValue = NSValue(cgPoint: toPoint)
        shake.fromValue = fromValue
        shake.toValue = toValue
        layer.add(shake, forKey: "position")
    }
    
    ///Set Background Color for various UIButton states.
    func setHHTabBackgroundColor(color: UIColor, forState: UIControl.State) {
        UIGraphicsBeginImageContext(CGSize(width: 1, height: 1))
        UIGraphicsGetCurrentContext()!.setFillColor(color.cgColor)
        UIGraphicsGetCurrentContext()!.fill(CGRect(x: 0, y: 0, width: 1, height: 1))
        let colorImage = UIGraphicsGetImageFromCurrentImageContext()
        UIGraphicsEndImageContext()
        self.setBackgroundImage(colorImage, for: forState)
        self.setBackgroundImage(colorImage, for: .highlighted)
    }
}


HHTabBarView/HHTabBarView/AppDelegate.swift
//
//  AppDelegate.swift
//  HHTabBarView
//
//  Created by Hemang Shah on 7/17/17.
//  Copyright © 2017 Hemang Shah. All rights reserved.
//

import UIKit

@UIApplicationMain
class AppDelegate: UIResponder, UIApplicationDelegate {

    var window: UIWindow?

    /////Start ------ Setup HHTabBarView ------
    
    //1
    //Initialize and keeping reference of HHTabBarView.
    let hhTabBarView = HHTabBarView.shared
    
    //Keeping reference of iOS default UITabBarController.
    let referenceUITabBarController = HHTabBarView.shared.referenceUITabBarController
    
    //2
    func setupReferenceUITabBarController() {
        
        //Creating a storyboard reference
        let storyboard = UIStoryboard(name: "Main", bundle: Bundle.main)
        
        //First and Second Tab ViewControllers will be taken from the UIStoryBoard
        //Creating navigation controller for navigation inside the first tab.
        let navigationController1: UINavigationController = UINavigationController(rootViewController: storyboard.instantiateViewController(withIdentifier: "FirstViewControllerID"))
        
        //Creating navigation controller for navigation inside the second tab.
        let navigationController2: UINavigationController = UINavigationController(rootViewController: storyboard.instantiateViewController(withIdentifier: "SecondViewControllerID"))
        
        //Third, Fourth and Fifth will be created runtime.
        let sofaViewController = UIViewController()
        sofaViewController.title = "Sofa"
        sofaViewController.view.backgroundColor = .white
        let sofaNavigationController: UINavigationController = UINavigationController(rootViewController: sofaViewController)
        
        let targetViewController = UIViewController()
        targetViewController.title = "Target"
        targetViewController.view.backgroundColor = .white
        let targetNavigationController: UINavigationController = UINavigationController(rootViewController: targetViewController)
        
        let umbrellaViewController = UIViewController()
        umbrellaViewController.title = "Umbrella"
        umbrellaViewController.view.backgroundColor = .white
        let umbrellaNavigationController: UINavigationController = UINavigationController(rootViewController: umbrellaViewController)
        
        //Update referenced TabbarController with your viewcontrollers
        referenceUITabBarController.setViewControllers([navigationController1, navigationController2, sofaNavigationController, targetNavigationController, umbrellaNavigationController], animated: false)
    }
    
    //3
    func setupHHTabBarView() {
        
        //Default & Selected Background Color
        let defaultTabColor = UIColor.black.withAlphaComponent(0.8)
        let selectedTabColor = UIColor(red: 28/255, green: 158/255, blue: 247/255, alpha: 1.0)
        let tabFont = UIFont.init(name: "Helvetica-Light", size: 12.0)
        let spacing: CGFloat = 3.0
        
        //Create Custom Tabs
        //Note: As tabs are subclassed of UIButton so you can modify it as much as possible.
        
        let titles = ["Location", "Refresh", "Sofa", "Target", "Umbrella"]
        let icons = [UIImage(named: "location-pin")!, UIImage(named: "refresh")!, UIImage(named: "sofa")!, UIImage(named: "target")!, UIImage(named: "umbrella")!]
        var tabs = [HHTabButton]()
        
        for index in 0...4 {
            let tab = HHTabButton(withTitle: titles[index], tabImage: icons[index], index: index)
            tab.titleLabel?.font = tabFont
            tab.setHHTabBackgroundColor(color: defaultTabColor, forState: .normal)
            tab.setHHTabBackgroundColor(color: selectedTabColor, forState: .selected)
            tab.imageToTitleSpacing = spacing
            tab.imageVerticalAlignment = .top
            tab.imageHorizontalAlignment = .center
            tabs.append(tab)
        }
        
        //Set HHTabBarView position.
        hhTabBarView.tabBarViewPosition = .bottom
        
        //Set this value according to your UI requirements.
        hhTabBarView.tabBarViewTopPositionValue = 64

        //Set Default Index for HHTabBarView.
        hhTabBarView.tabBarTabs = tabs
        
        // To modify badge label.
        // Note: You should only modify badgeLabel after assigning tabs array.
        // Example:
        //t1.badgeLabel?.backgroundColor = .white
        //t1.badgeLabel?.textColor = selectedTabColor
        //t1.badgeLabel?.font = UIFont.boldSystemFont(ofSize: 20.0)
        
        //Handle Tab Change Event
        hhTabBarView.defaultIndex = 0
        
        //Show Animation on Switching Tabs
        hhTabBarView.tabChangeAnimationType = .none                
        
        //Handle Tab Changes
        hhTabBarView.onTabTapped = { (tabIndex, isSameTab, controller) in
            if isSameTab {
                if let navcon = controller as? UINavigationController {
                    navcon.popToRootViewController(animated: true)
                } else if let vc = controller as? UIViewController {
                    vc.navigationController?.popToRootViewController(animated: true)
                }
            }
            print("Selected Tab Index:\(tabIndex)")
        }
    }
    
    //4
    // MARK: App Life Cycle
    func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]?) -> Bool {
        
        //Setup HHTabBarView
        setupReferenceUITabBarController()
        setupHHTabBarView()
        
        //This is important.
        //Setup Application Window
        window = UIWindow.init(frame: UIScreen.main.bounds)
        window?.rootViewController = referenceUITabBarController
        window?.makeKeyAndVisible()
        
        return true
    }

    /////End ------ SETUP HHTabBarView ------
    
    
    func applicationWillResignActive(_ application: UIApplication) { }
    func applicationDidEnterBackground(_ application: UIApplication) { }
    func applicationWillEnterForeground(_ application: UIApplication) { }
    func applicationDidBecomeActive(_ application: UIApplication) { }
    func applicationWillTerminate(_ application: UIApplication) {  }
}
